server/src/main/java/org/opensearch/index/codec/composite/Composite99Codec.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.codec.composite;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.codecs.Codec;
+import org.apache.lucene.codecs.DocValuesFormat;
+import org.apache.lucene.codecs.FilterCodec;
+import org.apache.lucene.codecs.lucene99.Lucene99Codec;
+import org.opensearch.common.annotation.ExperimentalApi;
+import org.opensearch.index.codec.PerFieldMappingPostingFormatCodec;
+import org.opensearch.index.mapper.MapperService;
+
+/**
+ *  Extends the Codec to support new file formats for composite indices eg: star tree index
+ *  based on the mappings.
+ *
+ * @opensearch.experimental
+ */
+@ExperimentalApi
+public class Composite99Codec extends FilterCodec {
+    public static final String COMPOSITE_INDEX_CODEC_NAME = "Composite99Codec";
+    private final MapperService mapperService;
+
+    // needed for SPI - this is used in reader path
+    public Composite99Codec() {
+        this(COMPOSITE_INDEX_CODEC_NAME, new Lucene99Codec(), null);
+    }
+
+    public Composite99Codec(Lucene99Codec.Mode compressionMode, MapperService mapperService, Logger logger) {
```

Comment:
@bharath-techie @msfroh I think we have an issue here with `PerFieldMappingPostingFormatCodec`, as per invariants, `PerFieldMappingPostingFormatCodec` always uses the latest available codec

```

    static {
        assert Codec.forName(Lucene.LATEST_CODEC).getClass().isAssignableFrom(PerFieldMappingPostingFormatCodec.class)
            : "PerFieldMappingPostingFormatCodec must subclass the latest " + "lucene codec: " + Lucene.LATEST_CODEC;
    }
```

 but `Composite99Codec` is explicitly bound to Lucene 9.9 (the latest for the current Apache Lucene version). I think we need to make a changes here while it is not too late (Apache Lucene 9.12 comes with new codecs)

Reply:
@reta I believe that should be fine. 

Lets consider the case when we upgrade to Lucene-912.

* The parametrized constructor for Composite99Codec breaks at compile time as PerFieldMappingPostingFormatCodec will accept only Lucene-912 Compression mode and the parameteized constructor in Composite99Codec will hence be removed (This is expected and preferred as we won't use this codec for any writes after upgrade).
* For writes which have happened using Composite99Codec, the reads will use the SPI constructor which is hardcoded to point to Lucene99Codec and does not cause issues.
* The new writes will happen via Composite9-12Codec which is obtained through CompositeCodecFactory as the implementation will switch to use Composite9-12Codec. The parametrized constructor will now only need to be present in this new codec.

For reference, this is how the custom-codecs plugin handles this as well.